benchmark/benchmark-atomic.cpp

```diff
@@ -0,0 +1,254 @@
+    template<typename AtomicType>
+    RAJA_HOST_DEVICE AtomicType operator()(AtomicType* acc, const AtomicType val) const {
+        return RAJA::atomicAdd(Policy {}, acc, val);
+    }
+};
+
+template<typename Policy>
+struct AtomicMax {
+    template<typename AtomicType>
+    RAJA_HOST_DEVICE AtomicType operator()(AtomicType* acc, const AtomicType val) const {
+        return RAJA::atomicMax(Policy {}, acc, val);
+    }
+};
+
+/// ExecPolicy wrapper for OpenMP
+struct ExecPolicyOMP {
+    using policy = RAJA::omp_for_exec;
+    using sync = RAJA::policy::omp::omp_synchronize;
+    static std::string PolicyName() {
+        std::stringstream ss;
+        ss << "OpenMP execution";
+        return ss.str();
+    }
+};
+
+/// Functor wrapping the desul implementation.  Wrapping the desul call ensure an identical signature with
+/// RAJA's implementations.  Wrapping the call in an functor allows simple type deduction for printing
+/// from within the benchmark.
+template<typename T, typename DesulAtomicSignature<T>::type atomic_impl>
+struct atomicWrapperDesul {
+    /// Call operator overload template that allows invoking DESUL atomic with a (int*)(T*, T) signature
+    RAJA_DEVICE T operator()(T * acc, T value) const {
+        return atomic_impl(acc, value, raja_default_desul_order{},
+                        raja_default_desul_scope{});
+    }
+};
+
+template<typename T>
+class IsDesul : public std::false_type {};
+
+template<typename T, typename DesulAtomicSignature<T>::type atomic_impl>
+class IsDesul<atomicWrapperDesul<T, atomic_impl>> : public std::true_type {};
+
+template<typename AtomicImplType>
+std::string GetImplName (const AtomicImplType& impl) {
+    if (IsDesul<AtomicImplType>::value) {
+        return "Desul atomic";
+    } else {
+        return "RAJA atomic";
+    }
+}
+
+/// Helper to deduce if atomic implementation is using desul under the hood
+//bool isDesul(typename RajaAtomicSignature<AtomicType>::signature atomic_impl) {
+//    bool is_desul =
+//        atomic_impl == atomicWrapperDesul<AtomicType, typename GPUAtomic::policy, desul::atomic_fetch_add> ||
+//        atomic_impl == atomicWrapperDesul<AtomicType, typename GPUAtomic::policy, desul::atomic_fetch_sub> ||
+//        atomic_impl == atomicWrapperDesul<AtomicType, typename GPUAtomic::policy, desul::atomic_fetch_>
+//
+//}
+
+template <class ExecPolicy, typename AtomicType, bool test_array, typename AtomicImplType>
+void TimeAtomicOp( AtomicImplType atomic_impl, uint64_t num_iterations = 2, int array_size = 100) {
+    RAJA::Timer timer;
+
+    // Allocate memory
+    AtomicType* device_value = nullptr;
+    int len_array = test_array ? array_size : 1;
+    if (IsGPU<typename ExecPolicy::policy>::value) {
+        AllocateAtomicDevice(&device_value, len_array);
+    } else {
+        device_value = new AtomicType [len_array];
+    }
+
+    for (int i = 0; i < num_iterations; ++i) {
+        // Don't time the first kernel launch
+        if (i == 1) {
```

Comment:
I don't think we want this test in the timing loop. I would just launch a kernel before the loop to warm up, then the timing loop would only contain the code that needs testing.

Reply:
I added a dry run before printing any outputs of the benchmarks